packages/binarytree/src/binaryTree.ts

```diff
@@ -0,0 +1,707 @@
+        this.DEBUG &&
+          this.debug(`Suffix: ${suffix}; Value: ${value === null ? 'null' : bytesToHex(value)}`, [
+            'get',
+          ])
+        values.push(value)
+      }
+      return values
+    }
+
+    return []
+  }
+
+  /**
+   * Stores a given `value` at the given `key` or performs a deletion if `value` is null.
+   * @param stem - the stem (must be 31 bytes) to store the value at.
+   * @param suffixes - array of suffixes at which to store individual values.
+   * @param values - the value(s) to store (or null for deletion).
+   * @returns A Promise that resolves once the value is stored.
+   */
+  async put(stem: Uint8Array, suffixes: number[], values: (Uint8Array | null)[]): Promise<void> {
+    if (stem.length !== 31) throw new Error(`expected stem with length 31, got ${stem.length}`)
+    if (values.length > 0 && values.length !== suffixes.length)
+      throw new Error(
+        `expected number of values (${values.length}) to equal number of suffixes (${suffixes.length})`,
+      )
+
+    this.DEBUG && this.debug(`Stem: ${bytesToHex(stem)}`, ['put'])
+    const putStack: [Uint8Array, BinaryNode | null][] = []
+
+    // If the tree is empty, initialize it.
+    if (equalsBytes(this.root(), this.EMPTY_TREE_ROOT)) {
+      await this._createInitialNode(stem, suffixes, values)
+      return
+    }
+
+    // Find the path to the node (or the nearest node) for the given stem.
+    const foundPath = await this.findPath(stem)
+
+    // We should always at least get the root node back
+    if (foundPath.stack.length === 0) throw new Error(`Root node not found in trie`)
+
+    // Step 1) Create or update the stem node
+    let stemNode: StemBinaryNode
+    // If we found a stem node with the same stem, we'll update it.
+    if (
+      foundPath.node &&
+      isStemBinaryNode(foundPath.node) &&
+      equalsBytes(foundPath.node.stem, stem)
+    ) {
+      stemNode = foundPath.node
+    } else {
+      // Otherwise, we'll create a new stem node.
+      stemNode = StemBinaryNode.create(stem)
+      this.DEBUG && this.debug(`Creating new stem node for stem: ${bytesToHex(stem)}`, ['put'])
+    }
+
+    // Update the values in the stem node.
+    for (let i = 0; i < suffixes.length; i++) {
+      const suffix = suffixes[i]
+      const value = values[i]
+      stemNode.setValue(suffix, value)
+      this.DEBUG &&
+        this.debug(
+          `Setting value for suffix: ${suffix} to value: ${value instanceof Uint8Array ? bytesToHex(value) : value} at stem node with stem: ${bytesToHex(stem)}`,
+          ['put'],
+        )
+    }
+
+    // If all values are null then we treat this as a deletion.
+    if (stemNode.values.every((val) => val === null)) {
+      if (foundPath.node !== null) {
+        this.DEBUG && this.debug(`Deleting stem node for stem: ${bytesToHex(stem)}`, ['put'])
+        putStack.push([this.merkelize(stemNode), null])
+      } else {
+        return // nothing to delete
```

Comment:
We should check if it's a deletion first and only then handle creating or updating the stem node.